server/src/main/java/org/cloudfoundry/identity/uaa/client/VmidentityClientMetadataProvisioning.java

```diff
@@ -0,0 +1,144 @@
+            OIDCClient client = idmClient.getOIDCClient(tenant, clientId);
+            return mapOIDCClient(client, tenant);
+        } catch (Exception e) {
+            logger.error("Unable to retrieve UI details for client: " + clientId, e);
+            throw new VmidentityDataAccessException("Unable to retrieve UI details for client: " + clientId);
+        }
+    }
+
+    @Override
+    public ClientMetadata update(ClientMetadata resource) {
+        logger.debug("Updating metadata for client: " + resource.getClientId());
+
+        try {
+            String tenant = VmidentityUtils.getTenantName(idmClient);
+            OIDCClient client = idmClient.getOIDCClient(tenant, resource.getClientId());
+
+            OIDCClient.Builder builder = new OIDCClient.Builder(client);
+
+            Map<String, Object> additionalInformation = client.getAdditionalInformation();
+            if (additionalInformation == null) {
+                additionalInformation = new HashMap<String, Object>();
+            }
+            if (resource.getClientName() != null) {
+                additionalInformation.put(ClientConstants.CLIENT_NAME, resource.getClientName());
+            }
+
+            if (resource.isShowOnHomePage()) {
+                additionalInformation.put(CLIENT_METADATA_SHOW_ON_HOME_PAGE, resource.isShowOnHomePage());
+            }
+
+            if (resource.getAppLaunchUrl() != null) {
+                additionalInformation.put(CLIENT_METADATA_APP_LAUNCH_URL, resource.getAppLaunchUrl());
+            }
+
+            if (resource.getAppIcon() != null) {
+                additionalInformation.put(CLIENT_METADATA_APP_ICON, resource.getAppIcon());
+            }
+
+            if (!additionalInformation.isEmpty()) {
+                builder.additionalInformation(additionalInformation);
+            }
+
+            idmClient.setOIDCClient(tenant, builder.build());
+        } catch (Exception e) {
+            logger.error("Unable to update metadata for client: " + resource.getClientId(), e);
+            throw new VmidentityDataAccessException("Unable to update metadata for client: " + resource.getClientId());
+        }
+
+        return retrieve(resource.getClientId());
+    }
+
+    public static ClientMetadata mapOIDCClient(OIDCClient client, String tenant) {
+        ClientMetadata clientMetadata = new ClientMetadata();
+        clientMetadata.setClientId(client.getClientId());
+        clientMetadata.setIdentityZoneId(tenant); // TODO check if this should be tenant or the zone (only matters for default)
+
+        Map<String, Object> additionalInformation = client.getAdditionalInformation();
+        if (additionalInformation != null) {
+            Object showOnHomepageFromAddInfo = additionalInformation.get(CLIENT_METADATA_SHOW_ON_HOME_PAGE);
+            if ((showOnHomepageFromAddInfo instanceof Boolean && (Boolean) showOnHomepageFromAddInfo || "true".equals(showOnHomepageFromAddInfo))) {
```

Comment:
The only reason I didn't was to stay semantically consistent with the JDBC variant. I'm not sure why they don't also do the same thing, but I didn't want to create behavioral differences. This is the piece of code lifted from `MultitenantJdbcClientDetailsService`:

```
                    @SuppressWarnings("unchecked")
                    Map<String, Object> additionalInformation = JsonUtils.readValue(json, Map.class);
                    Object autoApprovedFromAddInfo = additionalInformation.remove(ClientConstants.AUTO_APPROVE);
                    details.setAdditionalInformation(additionalInformation);
                    if (autoApprovedFromAddInfo != null) {
                        if ((autoApprovedFromAddInfo instanceof Boolean && (Boolean) autoApprovedFromAddInfo || "true".equals(autoApprovedFromAddInfo))) {
                            autoApproveScopes.add("true");
                        } else if (autoApprovedFromAddInfo instanceof Collection<?>) {
                            @SuppressWarnings("unchecked")
                            Collection<? extends String> approvedScopes = (Collection<? extends String>) autoApprovedFromAddInfo;
                            autoApproveScopes.addAll(approvedScopes);
                        }
                    }
```